pennylane/_qubit_device.py

```diff
@@ -29,100 +30,155 @@
-            parameters = {}
+        self.check_validity(circuit.operations, circuit.observables)
 
-        self.check_validity(queue, observables)
-        self._op_queue = queue
-        self._obs_queue = observables
-        self._parameters = {}
-        self._parameters.update(parameters)
+        # apply all circuit operations
+        self.apply(circuit.operations, circuit.diagonalizing_gates)
 
-        with self.execution_context():
-            self.pre_apply()
+        # determine the wires that are measured by the circuit
+        self._wires_measured = QubitDevice.active_wires(circuit.observables)
 
-            for operation in queue:
-                # Pass instances directly
-                self.apply(operation)
+        # generate computational basis samples
+        if (not self.analytic) or circuit.is_sampled:
+            self.generate_samples()
 
-            self.rotate_basis(observables)
+        # compute the required statistics
+        results = self.statistics(circuit.observables)
 
-            self.post_apply()
+        # Ensures that a combination with sample does not put
+        # expvals and vars in superfluous arrays
+        sample_return_types = (obs.return_type is Sample for obs in circuit.observables)
+        if any(sample_return_types) and not all(sample_return_types):
+            return self._asarray(results, dtype="object")
 
-            self.pre_measure()
+        return self._asarray(results)
 
-            results = self.statistics(observables)
+    @abc.abstractmethod
+    def apply(self, operations, rotations=None, **kwargs):
+        """Apply quantum operations, and execute the quantum circuit.
 
-            self.post_measure()
+        This method recieves a list of quantum operations queued by the QNode,
+        and should be responsible for:
 
-            self._op_queue = None
-            self._obs_queue = None
-            self._parameters = None
+        * Constructing the quantum program
+        * (Optional) rotating the quantum circuit using the rotation
+          operations provided. This diagonalizes the circuit so that arbitrary
+          observables can be measured in the computational basis.
+        * Compile the circuit
+        * Execute the quantum circuit
 
-            # Ensures that a combination with sample does not put
-            # expvals and vars in superfluous arrays
-            sample_return_types = (obs.return_type is Sample for obs in observables)
-            if any(sample_return_types) and not all(sample_return_types):
-                return self._asarray(results, dtype="object")
+        Args:
+            operations (list[~.Operation]): operations to apply to the device
+            rotations (list[~.Operation]): operations that rotate the circuit
+                pre-measurement into the eigenbasis of the observables.
+        """
 
-            return self._asarray(results)
+    @staticmethod
+    def active_wires(operations):
```

Comment:
```suggestion
    def active_wires(operators):
```

Reply:
This suggestion applies to the ``Hermitian`` case, where the ``Observable`` is not an ``Operation`` (but it is an ``Operator``)

Reply:
done!